### PR TITLE
Use implicit-hie to generate hie-*.yaml

### PR DESCRIPTION
**Pull request by michaelpj**
`implcit-hie` is the project that HLS uses to generate a cradle for a
cabal project when you don't have a `hie.yaml`. You can also just use it
to generate literal `hie.yaml` files, which is what I've done here, for both
the Cabal and Stack versions.

This is therefore a halfway house towards not needing the
`hie-*.yaml` files: they're checked in, so we can edit them if necessary,
or see where the generated output goes wrong (I had to make a few changes
to `.cabal` files to work around bugs).

Becuase I think we might want to tweak the otuput, I'm not asserting that the
checked-in files are exactly the output of `gen-hie` in CI, but there is an `update-hie`
script that will regenerate them. So as long as we keep the structure
broadly the same, you should be able to update them by running `update-hie`
and doing a bit of judicious `git add -p`.

TODO: explain this in CONTRIBUTING.

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `$(nix-build default.nix -A pkgsLocal.updateMaterialized)` to update the materialized Nix files
   - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
   - [ ] `$(nix-shell shell.nix --run fix-stylish-haskell)` to fix any formatting issues
- If you changed any Purescript files:
   - [ ] `$(nix-shell shell.nix --run fix-purty)` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by michaelpj** [comment]
To elaborate on why I don't just want to delete the files and let `implicit-hie` do its thing from inside HLS: there is a bug in `implicit-hie` (https://github.com/Avi-D-coder/implicit-hie/issues/24, fixed in an unreleased version), where comments in `hs-source-dirs` got into the output. If we didn't have the output checked in, it would be much harder to notice that and some components might just mysteriously not work properly, or you'd have to dig in the HLS logs.

**Comment by reactormonk** [comment]
Might just need a few comments where we use plugins, as otherwise the IDE displays large red lines everywhere because plugins crash the IDE. Otherwise: <3.

**Comment by michaelpj** [comment]
> Might just need a few comments where we use plugins, as otherwise the IDE displays large red lines everywhere because plugins crash the IDE.

As of the newer versions of HLS it seems to work on those components for me now!

File: nix/pkgs/default.nix [hunk 1]
```diff
@@ -154,9 +156,9 @@ in
 {
   inherit sphinx-markdown-tables sphinxemoji sphinxcontrib-haddock;
   inherit nix-pre-commit-hooks nodejs-headers;
-  inherit haskell agdaPackages cabal-install stylish-haskell hlint haskell-language-server hie-bios;
+  inherit haskell agdaPackages cabal-install stylish-haskell hlint haskell-language-server hie-bios gen-hie;
   inherit purty purty-pre-commit purs spago;
-  inherit fixPurty fixStylishHaskell updateMaterialized updateMetadataSamples updateClientDeps;
+  inherit fixPurty fixStylishHaskell updateMaterialized updateHie updateMetadataSamples updateClientDeps;
```
Comment (gilligan) [inline nix/pkgs/default.nix:161]: this is really only a minor thing but i sometimes wonder about consistent casing:

- `gen-hie` and `update-hie`
- `genHie` and `updateHie` 

of course this also applies to other attributes all over the place.

File: nix/pkgs/fix-purty/default.nix [hunk 2]
```diff
@@ -1,7 +1,6 @@
-{ writeScriptBin, runtimeShell, git, fd, purty }:
+{ writeShellScriptBin, runtimeShell, git, fd, purty }:
```
Comment (gilligan) [inline nix/pkgs/fix-purty/default.nix:1]: Good drive-by fix :+1:

File: nix/pkgs/update-hie/default.nix [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{ writeShellScriptBin, gen-hie }:
+writeShellScriptBin "update-hie" ''
+  ${gen-hie}/bin/gen-hie --cabal > hie-cabal.yaml
```
Comment (gilligan) [inline nix/pkgs/update-hie/default.nix:3]: I haven't actually run this - does this actually print something "useful" ? Otherwise i think you might want to add something like `@echo Updating 'hie-cabal.yaml'` or so. Makes it easier to spot in longer outputs in case this maybe gets run as part of some script or so.

**Review by gilligan — APPROVED** [review]
Assuming this actually *works* (which i haven't tried :sweat_smile: ) this LGTM. Only commented some very minor issues

File: nix/pkgs/update-hie/default.nix (hunk 3 above)
Comment (michaelpj) [reply nix/pkgs/update-hie/default.nix:3]: None of the other scripts that update things print stuff :shrug: generally I think "silent unless there's a problem" is okay?

File: nix/pkgs/default.nix (hunk 1 above)
Comment (michaelpj) [reply nix/pkgs/default.nix:161]: Yeah I agree. I made it consistent with the other things for now.

Comment (michaelpj) [reply nix/pkgs/default.nix:161]: Consistent with the inconsistency of the other things, that is!
